### DISCUSSION TITLE
WebSocket progress

### DISCUSSION THREAD

**Discussion by ste93** [robotology/yarp#2755] [General]
### work done:

- websocket implementation inside yarp
- tried compiling yarp into javascript with emscripten
	code: 
	emcmake cmake .. -DYCM_DIR=/home/ste/ws/ycm/build -DSKIP_ACE=ON -DBUILD_SHARED_LIBS=OFF

	emmake make -j16 YARP_os

	/home/ste/programs/emsdk/upstream/emscripten/tools/webidl_binder ../Bottle.idl glue

	em++ -s WASM=0  -ferror-limit=0 -s LINKABLE=1 -s EXPORT_ALL=1 -s MODULARIZE=1 -I../src/libYARP_conf/src/ -Isrc/libYARP_conf/src/ -I../src/libYARP_os/src/ lib/libYARP_os.a lib/libYARP_priv_hmac.a ../bottle_idl.cpp -s STRICT=0 --bind -s LLD_REPORT_UNDEFINED --no-entry -o out.js -s EXPORT_NAME=pippo -s USE_PTHREADS  -s WASM_BIGINT

	the error returned is 
	
	wasm2js: /b/s/w/ir/cache/builder/emscripten-releases/binaryen/src/passes/I64ToI32Lowering.cpp:471: void wasm::I64ToI32Lowering::visitAtomicCmpxchg(wasm::AtomicCmpxchg *): Assertion `curr->type != Type::i64 && "64-bit AtomicCmpxchg not implemented"' failed.
    em++: error: '/home/ste/programs/emsdk/upstream/bin/wasm2js --emscripten lib.wasm --mvp-features --enable-threads --enable-mutable-globals --enable-bulk-memory --enable-sign-ext' failed (received SIGABRT (-6))

	I think the operation Cmpxchg is not implemented in 64 bits, also with lowering.
	
	the operation is in wasm (https://github.com/WebAssembly/binaryen) in file https://github.com/WebAssembly/binaryen/blob/main/src/passes/I64ToI32Lowering.cpp
	
	if you compile em++ with WASM=1 then you can reproduce the error with:
	wasm2js --emscripten lib.wasm --mvp-features --enable-threads --enable-mutable-globals --enable-bulk-memory --enable-sign-ext
	
	
 - a small javascript library (https://github.com/ste93/yarp-websocket-library) contains the websocket management together with a small bottle manager. 
  it is implemented only for bottle messages, other types of messages are not implemented (like images)_ but images can be   retrieved via image src="http://yarp http protocol"
  another functionality implemented is to query the server to get portnames.

**Comment by traversaro** [comment]
Fyi @nunoguedelha

**Comment by nunoguedelha** [comment]
Hello @ste93, I'm implementing for AMI-IIT lab a telemetry visualizer based on Node.js (https://github.com/ami-iit/yarp-openmct), which uses Yarp Javascript bindings and websocket support through https://github.com/robotology/yarp.js.

I'm happy to see that websocket support is making its way into Yarp native features. Is there a release plan for your implementation, or an ongoing PR?
Thanks.

**Comment by nunoguedelha** [comment]
I was looking around and remembered this: https://github.com/robotology/yarp/pull/2621. So this is the a further iteration following the initial one #2621 ?

**Comment by ste93** [comment]
hi @nunoguedelha, actually part of the websocket incorporated in yarp has been already released (see https://github.com/robotology/yarp/pull/2621), for the library part we are experiencing issues in porting yarp in javascript (as explained above) and so we are at a still point, there is no expected release date for now. You can find a part of the library here https://github.com/ste93/yarp-websocket-library but for now we're not olanning any release of it

**Comment by nunoguedelha** [comment]
Great! Thanks for the info. Regarding the porting of Yarp on JS, out of curiosity, is the implementation close to or based on what was done in https://github.com/robotology/yarp.js?

**Comment by ste93** [comment]
no the implementation is standalone since the other  if i'm not wrong requires a server to be started, this implementation has an idea to talk directly from a web page to yarp server without any external server and network.

**Comment by nunoguedelha** [comment]
Hi @ste93 , sorry I didn't have time to react sooner to your message...

I'm not sure what you mean by "external server". There are two distinct things covered by [**robotology/yarp.js**](https://github.com/robotology/yarp.js):
-  **Javascript/Yarp-C++ bindings:** The module implements Javascript/Yarp bindings allowing to perform the usual operations, e.g. open/close/connect/disconnect Yarp ports, within the Javascript runtime environment **Node.js**. This sub-module does not require any server.
- **Websocket bridge:** It also implements a `browserCommunicator` sub-module in the same environment, which is a bridge interfacing websocket messages/requests (open/close/connect/disconnect/read/write) with the Yarp bindings that run the respective Yarp functions in the Node.js environment. An _express_ HTTP server is used to process the websocket requests. but such server can run on the localhost which is a pointer to the internal address of the loopback virtual network adapter (doesn't require a network connection).

My initial question was about the **Javascript/Yarp-C++ bindings** sub-module. Whatever "browser communicator" (bridge) is already implemented on YARP, it could re-use the binding functions in that sub-module.

**Comment by ste93** [comment]
Hi @nunoguedelha, I was referring to external server as the express HTTP server, now websocket communication are integrated in yarp. 
I don't know how the library yarp.js works and also if it is compiled with the latest updates. The major problem that we were facing was that we wanted to compile bindings like python but we had problem compiling in webassembly since, with emscripten that is the compiler we used to generate webassembly, some machine instructions are not implemented. I don't know if the library in yarp.js is compiled with the latest version or not and also if can talk directly to yarp without passing through the web server. If you know more we can update also on teams

**Reply by nunoguedelha** [reply]
yes sure, I'de be happy to discuss about that with you. See you there!
